### DISCUSSION TITLE
[FEATURE] Add a <py-module> tag

### DISCUSSION THREAD

**Discussion by jim22k** [pyscript/pyscript#1170] [Proposals]
## Problem statement
- A new developer is learning about Python modules, code organization, local imports, etc
- Up to this point, they have used PyScript to show off their work as a single, stand-alone HTML file
- While attempting to refactor code into Python modules, they run into several road blocks
  - CORS won't allow local files to be loaded by PyScript
  - Solutions asking the new developer to host their code on GitHub or other platform is a barrier to learning

## Desired Solution
Add a `<py-module>` tag which is typically added to the HTML header.

This is similar to a `<py-script>` tag, but is meant to represent a Python module. It is available to other PyScript code without needing to be called out in `py-env`. However, just like any Python module, it must be imported to be used. And it respects Python namespaces (please, let's never replicate the global variable fiasco that is Javascript).

Having these modules separated out would allow for an easy translation for `python-cli` to bundle up a Python project, and also be able to go in reverse (from a single PyScript HTML file with py-modules back into a regular Python project).

As modules follow a folder structure, there needs to be a `path` attribute on the tag. So the modules might look something like:
```python
<py-module path="foo/utils.py">
from datetime import datetime as dt

def format_date(dt_, fmt = "%m/%d/%Y, %H:%M:%S"):
    return dt_.strftime(fmt)

def now(fmt = "%m/%d/%Y, %H:%M:%S"):
    return format_date(dt.now(), fmt)
</py-module>

<py-module path="bar/espresso.py">
from ..foo.utils import now

def is_espresso_time():
    return now()[-1] in ["0", "4", "8"]
</py-module>
```

And then in the body, we might have
```python
<py-script>
from .bar.espresso import is_espresso_time

if is_espresso_time():
    # do something here
</py-script>
```

**Comment by sugatoray** [comment]
You can do this under `<py-env></py-env>`.

[**EDIT**]: I realized later that you are asking for this to be a separate tag that allows loading python modules. 

### Example using `<py-env>` tag

```python
<html>
    <head>
      <link rel="stylesheet" href="https://pyscript.net/alpha/pyscript.css" />
      <script defer src="https://pyscript.net/alpha/pyscript.js"></script>
      <py-env>
        - numpy
        - matplotlib
        - paths:
          - /data.py
      </py-env>
    </head>

  <body>
    <h1>Let's plot random numbers</h1>
    <div id="pymodule-code"></div>
    <py-script output="pymodule-code">
import matplotlib.pyplot as plt
from data import make_x_and_y

x, y = make_x_and_y(n=1000)

fig, ax = plt.subplots()
ax.scatter(x, y)
fig
    </py-script>
  </body>
</html>

```

**Comment by jim22k** [comment]
@sugatoray It's not an issue of loading python modules, but rather an issue of where those modules are stored.

I am aware of the `py-env` tag and local paths. But that solution runs into my 3rd bullet point in the original post. Web browsers normally block attempts to load local files (for security purposes). And attempting to share an HTML file with local paths would fail because data.py is not included in the HTML. The main file would need to be hosted on a server where the relative link can be resolved.

Both of these issues are a barrier to someone learning to code via PyScript. My proposal for a new `py-module` tag is an attempt to solve these issues. I want a way for a single HTML file to contain a directory's worth of python files. I don't expect that to be the easiest way to view the code, but I could easily imagine an editor which collects the individual `py-module` tags, arranges them into a tree view based on the paths, and lets you edit the file. So it would feel like a folder and project view, but all modules are actually contained in a single, sharable HTML file.

**Comment by marimeireles** [comment]
Hey @jim22k I like the idea.
Would you mind explaining what do you mean with `data.py`?
Also, in which kind of cases your third point would happen?

Thank you for your input! :)

**Comment by ckavidas** [comment]
Tbh I would strongly advise never exploring the idea of loading files from the local system without direct user interaction, browsers are designed in this manner to protect users.

It is easy enough to run a local server for education purposes, python includes a http server module for this purpose and some IDEs even have local server abilities as well.

**Comment by jim22k** [comment]
> Would you mind explaining what do you mean with data.py?

In sugatoray's example, if someone emailed me that HTML file, it would fail because `data.py` is not included. Someone would have to zip up the HTML+data.py, email it to me, and I would unzip it to run.

> Also, in which kind of cases your third point would happen?

The above example with the zip file would cause my 3rd point to happen. I unzip the folder, double-click the HTML file and it fails with a CORS error. As ckavidas pointed out above, there is a reason browsers won't load load files.

> It is easy enough to run a local server for education purposes, python includes a http server module for this purpose and some IDEs even have local server abilities as well.

This is true, but limits the ability to share a Pyscript HTML file. The one coding it likely has Python installed and knows how to `python -m http.server`, but their friend that they email the file to likely doesn't. There should be no burden on the one receiving a Pyscript file to know any coding. They would ideally be able to double-click an attachment and have a running dashboard or animation.

And asking the author to stand up a web server or get it hosted adds a layer of burden that I think Pyscript is trying to avoid. At least, that is how I read Peter Wang's keynote. He wants this to be used as an introduction to programming.

The idea behind a `<py-module>` tag is that it would expand on what is possible for a Pyscript file that you email to someone and they double-click the attachment. It expands beyond a very simple script into a real project with modules and hierarchy, but still remaining fully self-contained inside a single HTML file.

**Comment by ckavidas** [comment]
To me it sounds like you are looking for a way to bypass the browser's built-in protections for the sake of "good intentions". 

It seems like the scenario is as such:

> `Person_A` wants to share a pyscript application (as a html file) with `Person_B` but does not want to store the entire python code inside the html file.
> `Person_A` would like to include a separate python script `app.py` which automagically is located and ready to be imported.

This seems like an extreme extreme edge case and the premise confuses me very much. If I wanted to teach someone how to write python code I would set them up with a Jupyter notebook environment either through one of the free notebook services (example: google colab) or by installing Jupyter on their system.

**Comment by jim22k** [comment]
> `Person_A` wants to share a pyscript application (as a html file) with `Person_B` but does not want to store the entire python code inside the html file.
> `Person_A` would like to include a separate python script `app.py` which automagically is located and ready to be imported.

No, this is exactly **not** what I want to do. Instead, I very much want to store the entire python code inside the html file. However, I don't want a single, giant `<py-script>` tag with all my code inside. Python has modules for a reason. Once your code gets large enough, you refactor and split things into classes and modules. I want to allow something similar, but still contained within a single HTML file.

As for teaching with Jupyter notebooks, I agree that a lot of teaching can happen that way. But Pyscript already allows most of that functionality with `<py-repl>`. However, Jupyter notebooks can't teach relative imports and `__init__.py`.

**Comment by sugatoray** [comment]
Hi @jim22k,

> However, Jupyter notebooks can't teach relative imports and __init__.py.

I would disagree with this. You can write any file from jupyter notebooks. Save the contents of a cell as a file with `%%writefile <YOUR_INTENDED_FILEPATH>` at the top of that cell. I have done this even as an experiment to make contributions to some GitHub repositories.

It's not ideal. But if you want to teach how to create python modules, it works. I used Google Colab.

:fire: See an example here: 
- [demo: **An example of creating a python module in Google Colab**](https://github.com/sugatoray/code-share/blob/master/notebooks/demo_making_python_modules.ipynb)
